File: study/lua/src/app/ui/component/CommonTabIcon.lua
--
-- Author: Liangxu
-- Date: 2017-11-29
-- 侧边玉牌
local CommonTabIcon = class("CommonTabIcon")

local EXPORTED_METHODS = {
	"updateUI",
	"setCallback",
	"showRedPoint",
    "setSelected",
    "setNodeVisible"
}

function CommonTabIcon:ctor()
	self._target = nil
	self._imageBg = nil
	self._text = nil
	self._imageRedPoint = nil
	self._panelTouch = nil

	self._isOpen = nil
	self._index = nil
	self._callback = nil
end

function CommonTabIcon:_init()
	self._imageBg = ccui.Helper:seekNodeByName(self._target, "ImageBg")
	self._text = ccui.Helper:seekNodeByName(self._target, "Text")
	self._imageRedPoint = ccui.Helper:seekNodeByName(self._target, "ImageRedPoint")
	self._imageRedPoint:setVisible(false)
	self._panelTouch = ccui.Helper:seekNodeByName(self._target, "PanelTouch")
	self._panelTouch:setSwallowTouches(false)
	self._panelTouch:addClickEventListenerEx(handler(self, self._onPanelTouch))
end

function CommonTabIcon:bind(target)
	self._target = target
	self:_init()
	cc.setmethods(target, self, EXPORTED_METHODS)
end

function CommonTabIcon:unbind(target)
	cc.unsetmethods(target, EXPORTED_METHODS)
end

function CommonTabIcon:updateUI(txt, isOpen, index)
	self._isOpen = isOpen
	self._index = index

	local res = isOpen and Path.getUICommon("img_btn_tab01_nml") or Path.getUICommon("img_btn_tab01_dis")
	local color = isOpen and Colors.TAB_ICON_NORMAL or Colors.TAB_ICON_DISABLE
	self._imageBg:loadTexture(res)
	self._text:setString(txt)
	self._text:setColor(color)
end

function CommonTabIcon:setCallback(callback)
	self._callback = callback
end

function CommonTabIcon:showRedPoint(show)
	self._imageRedPoint:setVisible(show)
end

function CommonTabIcon:setSelected(selected)
	if not self._isOpen then
		return
	end

	local res = selected and Path.getUICommon("img_btn_tab01_down") or Path.getUICommon("img_btn_tab01_nml")
	local color = selected and Colors.TAB_ICON_SELECTED or Colors.TAB_ICON_NORMAL
	self._imageBg:loadTexture(res)
	self._text:setColor(color)
end

function CommonTabIcon:setNodeVisible(isVisible)
    self._imageBg:setVisible(isVisible)
    self._text:setVisible(isVisible)
    self._imageRedPoint:setVisible(isVisible)
    self._panelTouch:setVisible(isVisible)
end

function CommonTabIcon:_onPanelTouch(sender, state)
	if not self._isOpen then
		return
	end
	local offsetX = math.abs(sender:getTouchEndPosition().x - sender:getTouchBeganPosition().x)
	local offsetY = math.abs(sender:getTouchEndPosition().y - sender:getTouchBeganPosition().y)
	if offsetX < 20 and offsetY < 20  then
		self:setSelected(true)
		if self._callback then
			self._callback(self._index)
		end
	end
end

return CommonTabIcon

File: study/ts/script/ui/component/CommonTabIcon.ts
import { Path } from "../../utils/Path";

import { Colors } from "../../init";
import UIHelper from "../../utils/UIHelper";

const { ccclass, property } = cc._decorator;

@ccclass
export default class CommonTabIcon extends cc.Component {

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageBg: cc.Sprite = null;

    @property({
        type: cc.Label,
        visible: true
    })
    _text: cc.Label = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageRedPoint: cc.Sprite = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _panelTouch: cc.Node = null;
    _isOpen: any;
    _index: any;
    _callback: any;

    onLoad() {
        this._imageRedPoint.node.active = (false);
        var btn: cc.Button = this._panelTouch.addComponent(cc.Button);
        var clickEventHandler = new cc.Component.EventHandler();
        clickEventHandler.target = this.node; // 这个 node 节点是你的事件处理代码组件所属的节点
        clickEventHandler.component = "CommonTabIcon";// 这个是代码文件名
        clickEventHandler.handler = "_onPanelTouch";
        btn.clickEvents.push(clickEventHandler);
    }

    updateUI(txt, isOpen, index) {
        this._isOpen = isOpen;
        this._index = index;
        var res = isOpen ? Path.getUICommon('img_btn_tab01_nml') : Path.getUICommon('img_btn_tab01_dis');
        var color = isOpen ? Colors.TAB_ICON_NORMAL : Colors.TAB_ICON_DISABLE;
        UIHelper.loadTexture(this._imageBg, res);
        this._text.string = (txt);
        this._text.node.color = (color);
    }

    setCallback(callback) {
        this._callback = callback;
    }

    showRedPoint(show) {
        this._imageRedPoint.node.active = (show);
    }

    setSelected(selected) {
        if (!this._isOpen) {
            return;
        }
        var res = selected && Path.getUICommon('img_btn_tab01_down') || Path.getUICommon('img_btn_tab01_nml');
        var color = selected && Colors.TAB_ICON_SELECTED || Colors.TAB_ICON_NORMAL;
        UIHelper.loadTexture(this._imageBg, res);
        this._text.node.color = (color);
    }

    _onPanelTouch(sender, state) {
        this.setSelected(true);
        if (this._callback) {
            this._callback(this._index);
        }
    }

    setVisible(v) {
        this.node.active = v;
    }

}
